0000-encryption-api/README.md

```diff
@@ -0,0 +1,135 @@
+
+[Libsodium](https://libsodium.gitbook.io/doc/) for example uses these two types as public-key encryption and secret key encryption.
+
+Public key encryption is generally applied for key exchange or secret key wrapping. It requires two different keys–*public* and *private*, linked together with a one way mathematical
+relationship. The *private* key is used to decrypt a message and cannot be recovered from the *public* key and MUST remain private. The *public* key is used to encrypt a message.
+
+Secret key encryption uses the same key to encrypt and decrypt. It is only shared among all involved parties as a *shared secret key*.
+
+Users will create keys using one of these two types and pass them to the appropriate method.
+Rust typing will prevent a user from passing a secret key to a public key encryption method and visa versa.
+
+Public-Key encryption methods MUST use the naming convention **ALGORITHM or CURVE{SIZE}-SCHEME**. Below are some examples
+that could be implemented:
+
+- Rsa2048Oaep
+    - Encryption with an RSA 2048 bit public key using [Optimal Asymmetric Encryption](https://tools.ietf.org/html/rfc8017#section-7.1).
+    - Decryption with an RSA 2048 bit private key.
+    - This method does not use a MAC or authentication tag.
+- X25519Chacha20Poly1305
+    - Encryption with shared secret derived from X25519 public key and ephemeral X25519 private key using CHACHA-POLY1305.
+    - Decryption with shared secret derived from X25519 private key and ephemeral X25519 public key using CHACHA-POLY1305.
+    - Uses authentication tag.
+
+Secret-key encryption methods MUST use the naming convention **ALGORITHM{SIZE}{-MAC}**. Below are some examples
+that could be implemented:
+
+- Aes256Gcm
+- Chacha20Poly1305
+- Twofish256CbcHmacSha2_256
+- Blowfish448CbcHmacSha2_256
+
+Secret-key encryption methods require the use of *nonces* or *initialization vectors (iv)*, the two terms are often used interchangeably.
+We use the term *nonce* here. Nonces must be unique for a given key and message pair to be prevent certain types of attacks.
+To simplify the interfaces for Ursa users, *nonce* generation will be handled by Ursa and not a required concern for developers.
+
+Authenticated encryption methods are strongly preferred and any other methods should be considered hazardous materials–only used by those who know what they are doing.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+All proposed code will be implemented in **libursa**.
+**Libursa** will use existing traits and structures where possible.
+`KeyGenOption`, `PublicKey`, `PrivateKey`, `SecretKey`, and `MacKey`
+are already implemented.
+
+Rust structs and traits will look and work like so:
+```rust
+pub trait PublicKeyEncryptionScheme {
+    fn new() -> Self;
+    fn keypair(&self, options: Option<KeyGenOption>) -> Result<(PublicKey, PrivateKey), CryptoError>;
```

Comment:
General comment:  I agree with Lovesh here.  We should make naming conventions as explicit as possible.  So this should be something like "Encrypt-PKE" to specify that it is the encryption algorithm of a public key  scheme.

This would extend to other functions as well, including KeyGen (we may need multiple KeyGen algorithms in the future, including something like KeyDerive).

Reply:
Which is why I'm suggesting we enumerations so other methods can be added later without breaking the API.

Reply:
The other way to do this is like this:
```rust
pub trait KeyPair<P, S>
where
    P: PublicKey,
    S: SecretKey {
    type Error;
    fn generate(options: Option<KeyGenOptions>) -> Result<Self, Self::Error>;
}
````
I think it gives you more flexibility and allows for a nicer client experience using the ? operator:
```rust
let keypair = Ed25519KeyPair::generate(Ed25519Defaults)?;
```
If the KeyGenOptions is also a trait along with all of the other "options" types passed into API functions then a particular implementation--in this case Ed25519--could have just a single type that implements all of the *Options traits:
```rust
trait KeyGenOptions {
    static KEYBYTES: usize;
}
trait NonceGenOptions {
    static NONCEBYTES: usize;
}

struct Ed25519DefaultOptions;

impl KeyGenOptions for Ed25519DefaultOptions {
    static KEYBYTES: usize = 32;
}

impl NonceGenOptions for Ed25519DefaultOptions {
    static NONCEBYTES: usize = 24;
}
```
Let's do Rust harder. I like where you're going with this but we should be coding in terms of traits more I think.

Reply:
+1 to @dhuseby's comment

Reply:
Yes I want traits as well so we can rapidly create types like 

pub type Aes256Gcm = Gcm<Aes256>;